### PR TITLE
New lens distortion calibration software

### PR DESCRIPTION
**Pull request by fabiencastan**
## Description

New software to calibrate the lens distortion parameters from an image of a checkerboard.

### CODE REVIEW

File: src/aliceVision/sfm/BundleAdjustmentCeres.cpp [hunk 1]
```diff
@@ -30,6 +30,154 @@ namespace sfm {
...
+    {
+      if (_lockFocalRatio)
+      {
+        x_plus_delta[0] = x[0] + delta[posDelta]; 
+        x_plus_delta[1] = x[1] + _focalRatio * delta[posDelta];
+        posDelta++;
+      }
+      else
+      {
+        x_plus_delta[1] = x[1] + delta[posDelta];
```
Comment (fabiencastan) [inline src/aliceVision/sfm/BundleAdjustmentCeres.cpp:87]: ``
```suggestion
        x_plus_delta[0] = x[0] + delta[posDelta];
```

File: src/aliceVision/sfm/BundleAdjustmentCeres.cpp [hunk 2]
```diff
@@ -30,6 +30,154 @@ namespace sfm {
...
+    _focalRatio(focalRatio),
+    _lockFocal(lockFocal),
+    _lockFocalRatio(lockFocalRatio),
+    _lockCenter(lockCenter),
+    _lockDistortion(lockDistortion)
+  {
+    _distortionSize = _globalSize - 4;
+
+    std::cout << "????????" << _focalRatio << std::endl;
+
```
Comment (fabiencastan) [inline src/aliceVision/sfm/BundleAdjustmentCeres.cpp:46]:
```suggestion

    std::cout << "????????" << _focalRatio << std::endl;

```

**Review by simogasp — CHANGES_REQUESTED** [review]
I did a pass in the calibration module.
It's very hard to process these kinds of PRs, it's too big, even the browser struggles.
It is better to have PRs that are self-contained and focusing on a single feature. Here we have calibration, a new camera model and god knows what else.
If we want to keep the codebase in check we should avoid all this.
Just my thought.

File: src/aliceVision/calibration/checkerDetector.cpp [hunk 3]
```diff
@@ -0,0 +1,1338 @@
...
+    while (candidates.size() > 0)
+    {
+        //Get the worst candidate and remove it
+        IndexT j = candidates.back().col;
+        IndexT i = candidates.back().row;
+        candidates.pop_back();
+        board(i, j) = UndefinedIndexT;
+
+        //If the next candidate is very bad, continue
+        if (candidates.back().score > 1e6)
```
Comment (fabiencastan) [inline src/aliceVision/calibration/checkerDetector.cpp:802]: I got a crash on `candidates.back()` as there is a pop_back just before.

File: src/aliceVision/camera/DistortionRadial.hpp [hunk 4]
```diff
@@ -122,11 +122,7 @@ class DistortionRadialK1 : public Distortion {
       return Eigen::Matrix<double, 2, 1>::Zero();
     }
 
-//    const double r2 = r * r;
-    
-    /*const double r_coeff = 1.0 + k1 * r2;*/
-
-    const Eigen::MatrixXd ret = p * 2.0 * k1 * r;
+    const Eigen::MatrixXd ret = p * r * r;
```
Comment (fabiencastan) [inline src/aliceVision/camera/DistortionRadial.hpp:125]: k1 is not used?

File: src/software/utils/main_sfmTransfer.cpp [hunk 5]
```diff
@@ -221,11 +221,13 @@ int aliceVision_main(int argc, char **argv)
 
     if (matchingMethod == EMatchingMethod::FROM_INTRINSICID) 
     {
-        for (auto intrinsic : sfmData.getIntrinsics()) 
+        for (auto & intrinsic : sfmData.getIntrinsics())
```
Comment (fabiencastan) [inline src/software/utils/main_sfmTransfer.cpp:224]: the std::pair of the map is temporary.

**Pull request merged**